share/doc/revng/references/abi-definition.md

```diff
@@ -0,0 +1,213 @@
+
+## `BigArgumentsUsePointersToCopy` (`bool`)
+
+This allows a non-position-based ABI to take advantage of one of the main traits of position-based ABI: ability to use pointers to copy to pass big aggregates.
+
+This is one of the defining features of arm64 ABI, which allows it to get the best of both approaches - an ability to pass aggregates by copy in registers from the non-position-based approach while also being able to sometimes avoid a copy of a bigger aggregate from the position-based approach.
+
+## `NoRegisterArgumentsCanComeAfterStackOnes` (`bool`)
+
+States whether ABI allows a stack argument (mainly if it's too big to be placed in the registers) to precede other register arguments.
+
+For example, if there is a `Big` struct that has to use the stack, and
+a function like
+```cpp
+void function(Big, uint32_t);
+```
+if this value is set to true, both argument will be passed on stack, otherwise, only the struct will.
+
+## `AllowUnnaturallyAlignedTypesInRegisters` (`bool`)
+
+States whether ABI allows unnaturally aligned types[^1] to be passed in registers. When set to false, all the unnatural types will only be put on stack.
+
+## `CalleeIsResponsibleForStackCleanup` (`bool`)
+
+Specifies who is responsible for cleaning the stack after the function call. If equal to `true`, it's the callee, otherwise it the caller.
+
+## `FloatsUseGPRs` (`bool`)
+
+Setting this to true disables usage of vector registers entirely.
+
+## `StackArgumentsUseRegularStructAlignmentRules` (`bool`)
+
+Setting this to true allows stack arguments to use slots smaller than a word (makes them behave like a single struct instead of a bunch of arguments): so they don't get extended unnecessarily. A notable ABI using this is apple flavor of the aarch64.
+
+For example
+```cpp
+void function(/* enough arguments of all the registers */, uint8_t, uint8_t);
+```
+When this is set to true, both `uint8_t` will get placed as if they were a struct, that is without any padding in-between. When this is false, both of them will get extended to register size before getting placed, leading to padding.
+
+## `UseStrictAggregateAlignmentRules` (`bool`)
+
+Treat all the structs passed using the stack as if they were aligned unnaturally[^1].
+
+This means that they are going to handled with extreme care (which usually just means extra padding).
+
+## `StackAlignment` (`unsigned`)
+
+States the required alignment of the stack at the point of a function call in bytes.
+
+> **_Note:_** states minimum value for ABIs supporting multiple different stack alignment values, for example, if the ABI requires the stack to be aligned on 4 bytes for internal calls but on 8 bytes for interfaces (like 32-bit ARM ABI), the value of `StackAlignment` should be equal to 4.
```

Comment:
> the ABI requires the stack to be aligned on 4 bytes for internal calls but on 8 bytes for interfaces

WHAT?
What is an "internal call"?
ABI are evil.

Reply:
Yeah, this is courtesy of arms. They call it "public interface"s and "private interface"s though. Maybe we should switch to the same terminology 🤔

Reply:
Wait but is this private vs public something under the control of the dev?
Is this related to how far the target function is?
Or maybe whether is static or not?
Or maybe its visibility?

I need some reference.

Reply:
Ok, digging a bit deeper into it, it doesn't seem as bad.

This is what their docs have to say on the matter:
```
This definition of conformance gives maximum freedom to implementers. For example, if it is
known that both sides of an externally visible interface will be compiled by the same compiler,
and that the interface will not be publicly visible, the AAPCS64 permits the use of private
arrangements across the interface such as using additional argument registers or passing data
in non-standard formats. Stack invariants must, nevertheless, be preserved because an
AAPCS64-conforming routine elsewhere in the call chain might otherwise fail. Rules for use of
IP0 and IP1 must be obeyed or a static linker might generate a non-functioning executable
program.
Conformance at a publicly visible interface does not depend on what happens behind that
interface. Thus, for example, a tree of non-public, non-conforming calls can conform because
the root of the tree offers a publicly visible, conforming interface and the other constraints are satisfied.
```

To me it just sounds like "hey, you can do what you want with static (or maybe even partially inlined?) calls, but the rules for public interfaces are as follows".

Reply:
Please use quote (`> some text here`) for long pasting text. Using code snippets disable word-wrapping and makes it hard to read.

> To me it just sounds like "hey, you can do what you want with static (or maybe even partially inlined?) calls, but the rules for public interfaces are as follows".

Same. So... can we drop this option?

Reply:
Option, no - we still need to know alingment 😆.

Note - probably 🤔.

Reply:
Ah, but wait... It's currently set to the less restrictive value 🤔.

This might be tricky. Because I don't think I would have set to the more obscure value without a good reason. Some static function somewhere was probably interfering. And I'm not sure we're ready to split `arm-aapcs-static` ABI from `arm-aapcs` 😭.

Reply:
I'm confused, the whole ABI project should entirely ignore optimizations that the compiler can do for `static` functions. This is very important: handling what the compilers do for `static` functions is a non-goal that should not be pulled in. Maybe one day we'll add `-static` ABIs, but not today.

We should entirely ignore the whole paragraph you posted. Compilers are already doing whatever they want in that case, there's no need for permission by the ABI.

Reply:
> Compilers are already doing whatever they want in that case, there's no need for permission by the ABI.

Well, turns out, they do in fact have a permission! Who would have guessed.

> This is very important: handling what the compilers do for `static` functions is a non-goal that should not be pulled in.

But here the catch: we can't reliably tell if what we are looking at is a static function or not. So we still need at least some basic handling.

Let's look at this specific case. Arm abi requires that after every instruction, the stack is aligned at 4 bytes. But, on a "public interface" instruction, it requires 8 byte alignment.

By encoding less strict one in our system, we prevent horrible breakage in the case we run into static function (memory overlap), while only sacrificing trailing padding, which, let's be honest, we don't even care about that much - we already take some liberties around last argument as is.

Reply:
We're trying to guess what the compiler does in case of `static` functions, which is a losing battle.

> we prevent horrible breakage in the case we run into static function

Can you make an example of the consequence of encoding StackAlignment to 8 in such a situation? It's not clear to me.
The fact that this actually happens and has catastrophic consequence that we cannot prevent needs some significant backing evidence. Otherwise, I really want to strictly follow the ABI rules, and the ABI says 8.
If we were a compiler and we used this information to emit a call, we'd lose the alignment constraint and emit invalid code.

But in any case, if something horrible happens when we someone does not respect the ABI, that's fine. The problem is that conversion from RFT to CABI should have failed. If we don't have enough information to make it fail, it might be reasonable to add to RFT a new field to detect such situations. For instance, we could add `StackAlignmentUponCall` and have `detect-abi` try to detect it (maybe).

Reply:
> Can you make an example of the consequence of encoding `StackAlignment` to 8 in such a situation?

It leads to the layout reporting arguments as too big. As in we would grab some memory that is not an argument, and label it as such. I'd say it's pretty likely that there could be a local variable or something in there - so now we observe extra reads and writes into such an "argument".

Given, we'd needs to investigate to reliably tell what'd happen in our case because of such misreporting, depends on how much we rely on the layout information. That being said, consider the opposite error: we report too little padding, end up with some hole in memory (that was never accessed to begin with), which will probably just disappear at some point (when we are deciding what to promote into an variable?), nobody is hurt.

> If we were a compiler

If we were a compiler, we'd never be able to get away with a handful of options! We'd need something akin to the headache clang is encoding in their :cursed_parrot: code gen tables.

I get what you're trying to say, but also, I'm not sure what's more important: being closer to the source of truth (even though we're never matching it) or going up in flames as rarely as possible.

> The problem is that conversion from RFT to CABI should have failed.

Well, yeah, if we could do that, the world would be a much nicer place (I would even let you ride my unicorn! /j). The problem is that everything related to alignment we have is pretty much guesswork. There's usually no way to tell padding apart for unused words. And it sometimes even depends on types alone - and we obviously can't rely on the fact that we recovered those correctly.

Reply:
I still do not understand the extent of the problem. I need a specific example, in a specific scenario (e.g., RFT->CABI or CABI->Layout), with a specific prototype where you can say "if `StackAlignment` is 8 this happens, if it's 4 this other thing happens".

Reply:
Done, as per the mattermost discussion.